pipelines/shared/nextflow/color_workflow.nf

```diff
@@ -0,0 +1,202 @@
+
+process get_id_list {
+    publishDir params.final_output_dir, mode: 'copy'
+    input:
+        path cluster_id_map
+        path singletons
+        path seqid_source_map
+    output:
+        path 'id_lists/', emit: 'id_lists'
+        path 'cluster_sizes.txt', emit: 'cluster_sizes'
+    """
+    mkdir id_lists
+    mkdir id_lists/uniprot_ids id_lists/uniref90_ids id_lists/uniref50_ids
+    perl $projectDir/../shared/perl/get_id_lists.pl --cluster-map $cluster_id_map --singletons $singletons \
+        --uniprot id_lists/uniprot_ids --uniref90 id_lists/uniref90_ids --uniref50 id_lists/uniref50_ids \
+        --seqid-source-map $seqid_source_map --cluster-sizes cluster_sizes.txt \
+        --config ${params.efi_config} --db-name ${params.efi_db}
+    """
+}
+
+process get_fasta {
+    publishDir params.final_output_dir, mode: 'copy'
+    input:
+        path id_list_dir
+    output:
+        path 'fasta/'
+    """
+    mkdir fasta
+    mkdir fasta/uniprot_ids fasta/uniref90_ids fasta/uniref50_ids
+    for id_file in $id_list_dir/uniprot_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    for id_file in $id_list_dir/uniref90_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    for id_file in $id_list_dir/uniref50_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    """
```

Comment:
I would try to make `get_fasta` accept a tuple of the path to the idlist and the verison tag. In the `workflow` section, you could then filter for files that exist and aggregate by version tag and pass all of the id files for that version to the process or pass every pair of (file, version) to `get_fasta` to parallelize all of the lookups. Either way I would try to avoid the explicit for loops and manual directory management. I can elaborate on this if needed.

Reply:
https://www.nextflow.io/docs/latest/channel.html#operators